### ISSUE TITLE
Please let us use Ramda instead of proxies to change docs

### ISSUE THREAD

**Issue by bionicles** [automerge/automerge#294]
Dear Automerge Devs,

One thing I had issues with, and switched away from this platform over, was the requirement to use custom OOP / mutation syntax during the changes. It seems odd to force on users a specific API to mutate state, because that puts automerge devs who really care about the CRDT part, on the hook to re-invent the wheel of every possible way to mutate state. That's not even feasible and makes this project much harder due to exponential / infinite feature creep. It makes usage harder because we have to wonder if a certain function we would like to use on an object would be replicated in the automerge API, then find your docs on that, and you're on the hook to document the proxies. It's a lot of extra trouble all around.

I just find the proxies concept, frankly annoying, because automerge devs must re-implement proxy-everything, and automerge users are stuck on a subset of re-implemented re-documented proxy-everything methods.

Would it please be possible to de-spec / reduce the automerge project to literally delete the proxies and make it work more like Redux? 

In Redux, reducers take a state, and return a new state, done. There's no need for Redux devs to invent proxies for every data structure api in javascript, and there's no need for users to obey some custom / limited OOP syntax. Instead of this 'proxies.js' concept, what if there were simply a diff written as a change operation?

This enables automerge to be lighter / simpler / faster / less code / less work / more options / less issues / less bugs.

 I'd rather just use Ramda, return a brand new json object, and have you take a diff. Maybe that's a bit slower in terms of top-end performance, because of copying and diffs, but IMO it's way simpler to maintain and use going forward. 

What do you think? 

Thanks in advance for continued development of Automerge!

Bion

P.S. here's a cool diff function: https://gist.github.com/plukevdh/dec4b41d5b7d67f83be630afecee499e

**Comment by pvh** [comment]
Thanks for writing. 

I agree that the current automerge change API isn't ideal, but state diff not only has worse performance but it also provides less accurate merges. The current automerge interface keeps track of object identities as well as values which is particularly helpful for producing high fidelity array changes. I know this because we've implemented a [diff-based interface](https://github.com/inkandswitch/vscode-hypermerge/blob/master/src/Diff.ts) and @jeffpeterson had to spend a bunch of time dealing with bad changes generated by inaccurate heuristics.

All this said, the `change` function is relatively straightforward and replacing that part of the front-end with something more to your taste shouldn't be too difficult. What I would probably recommend would be a more explicit API to the underlying CRDT objects instead of the implicit "instrumented object" style, then I'd use something like the above linked `Diff.ts` to modify the objects automatically. This would give you the benefit of a more explicit API without imposing the performance costs and reduced fidelity of a diff-based system on all automerge consumers.

**Comment by pvh** [comment]
Oh, also -- while I wholeheartedly encourage you to adopt ramda, if you do decide to pursue this, it's an explicit strategy of automerge to minimize external dependencies so I'd suggest building a simple automerge-ramda-diff library for any glue code there.

**Comment by ept** [comment]
To expand on what @pvh said: it's not too difficult to add your own frontend API to Automerge, and hook it into the existing CRDT logic. I suggest looking at the [documentation for the change format](https://github.com/automerge/automerge/blob/3fc5fffd78b9afa4c9bf048b656b37345f97435e/BINARY_FORMAT.md#json-representation-of-changes) to get started. You can use any API you like, as long as it can produce changes in the format described here. (The documentation I linked to is for the performance branch, which is becoming an Automerge 1.0 release candidate. The format for the currently released version of Automerge is [slightly different](https://github.com/automerge/automerge/blob/0678f39eccb9043c5b0ec2189ead49115d88449c/INTERNALS.md#change-structure-and-operation-types).

Making Automerge work more like Redux has a problem: with Redux you can have arbitrary code in the reducer, and when you have several users collaborating, different users may apply updates in a different order. (For example, you make change A while I concurrently make change B. I then send my change to you, so you first apply A and then B. You send your change to me, so I first apply B and then A.) When A and B can be arbitrary code, there is no guarantee that executing first A and then B yields the same final result as first executing B and then A, so there is no consistency guarantee. Automerge can guarantee eventual consistency because it does not allow arbitrary reducer code, but rather limits a change to the types of operation that can be performed in the course of a change.

**Comment by echarles** [comment]
> Making Automerge work more like Redux has a problem:

@ept What do you think of https://github.com/local-first-web/state who describes it-self like `an automatically replicated Redux store that gives your app offline capabilities and secure peer-to-peer synchronization superpowers` developed by @HerbCaudill and using Automerge?

**Comment by ept** [comment]
@HerbCaudill's work is great! I recommend using it, or at least looking at it as a model for how to do things.

(Herb, I didn't realise you had renamed Cevitxe! Do you want to update the link in Automerge's README?)

**Comment by HerbCaudill** [comment]
> Making Automerge work more like Redux 
> 
> what do you think of local-first-web/state ... `an automatically replicated Redux store...

FWIW @local-first/state exposes a Redux store, but the reducers have to be written as Automerge change functions. So if the proxy change functions are the obstacle to using Automerge, then that library isn't the solution.

**Comment by bionicles** [comment]
@ept @HerbCaudill @echarles @pvh 

hey, could we reopen this issue to continue discussion about adding more functional programming stuff to automerge? I avoid slack cuz it's distracting, but git issues are nice for async communication!

would you all be openminded to copy a few cool functions from ramda to auto merge to make it a lot nicer for devs?

TLDR: nested CRUD ops - without the R, because we can just use R.path
1. `assocPath(steps, value, obj)` -> `obj[step1][step2]` = value (string steps create objects, integer steps create arrays)
2. `mergePath(steps, update, obj)` -> `obj[step1][step2]` = mergeDeepRight(obj[step1][step2], update)
3. `dissocPath(steps, obj)` -> `del obj[step1][step2]`

4. can we make a function to wrap other functions and add the change format stuff? then we could extend this with custom changes more easily

5. I also really `evolvePath = (steps, transformations, obj) => over(lensPath(steps), evolve(transformations), obj)` -> `obj[step1][step2] = evolve(transformations, obj[step1][step2]` cuz you can apply different functions to different fields of nested objects

6. non-path versions of the same are cool too for simpler stuff, but we could just pass an empty list or undefined for 'steps' in these cases

### could this stuff be implemented in a clean, extensible way with [optics / lenses](https://github.com/calmm-js/partial.lenses#---the-why-of-optics) ? perhaps Automerge Lens could handle change metadata whenever the setter is used. 

I'd love to see some simple examples of how to add [custom automerge functions](https://github.com/ramda/ramda/wiki/Cookbook) -- this could enable the community to extend the platform, and that makes folks more engaged and also makes automerge more useful

**Comment by pvh** [comment]
I think you'll find that replacing the Automerge frontend with something Ramda-based is probably a good way to get you what you're looking for. It's an intentionally narrow and simple API to the backend and designed to be replaceable. In fact, there are now two reasonably mature backends (JS and Rust) and several frontends in varying states of maturity (JS, Rust, and Python.)
